src/libvirt.c

```diff
@@ -0,0 +1,162 @@
+
+  size_t argc = 1;
+  napi_value args[1];
+  status = napi_get_cb_info(env, info, &argc, args, NULL, NULL);
+  assert(status == napi_ok);
+
+  if (argc > 1) {
+    napi_throw_type_error(env, NULL, "Wrong number of arguments");
+    return NULL;
+  }
+
+  status = napi_typeof(env, args[0], &valuetype);
+  assert(status == napi_ok);
+
+  if (valuetype != napi_undefined && valuetype != napi_string) {
+    napi_throw_type_error(env, NULL, "Wrong argument type");
+    return NULL;
+  }
+
+  char* name = NULL;
+
+  if(valuetype == napi_string) {
+    char buffer[128];
+    size_t buffer_size = 128;
+    size_t copied;
+    status = napi_get_value_string_utf8(env, args[0], buffer, buffer_size, &copied);
+    assert(status == napi_ok);
+    name = buffer;
+  }
+
+  virConnectPtr c_retval;
+  c_retval = virConnectOpen(name);
+
+  if(c_retval == NULL) {
+    status = napi_get_null(env, &n_retval);
+    assert(status == napi_ok);
+    return n_retval;
+  }
+
+  status = napi_create_external(env, c_retval, NULL, NULL, &n_retval);
+  assert(status == napi_ok);
+
+  return n_retval;
+}
+
+napi_value libvirt_virDomainLookupByName(napi_env env, napi_callback_info info) {
+  napi_status status;
+  napi_value n_retval;
+  napi_valuetype valuetype;
+
+  size_t argc = 2;
+  napi_value args[2];
+  status = napi_get_cb_info(env, info, &argc, args, NULL, NULL);
+  assert(status == napi_ok);
+
+  if (argc != 2) {
+    napi_throw_type_error(env, NULL, "Wrong number of arguments");
+    return NULL;
+  }
+
+  // arg0
+  status = napi_typeof(env, args[0], &valuetype);
+  assert(status == napi_ok);
+  if (valuetype != napi_external) {
+    napi_throw_type_error(env, NULL, "Wrong argument type");
+    return NULL;
+  }
+
+  virConnectPtr conn = NULL;
+  status = napi_get_value_external(env, args[0], (void**)&conn);
+  assert(status == napi_ok);
+  assert(conn);
+
+  // arg1
+  status = napi_typeof(env, args[1], &valuetype);
+  assert(status == napi_ok);
+
+  if (valuetype != napi_string) {
+    napi_throw_type_error(env, NULL, "Wrong argument type");
+    return NULL;
+  }
+
+  char name[128];
+  size_t name_size = 128;
+  size_t copied;
+  status = napi_get_value_string_utf8(env, args[1], name, name_size, &copied);
+  assert(status == napi_ok);
+
+  virDomainPtr c_retval;
+  c_retval = virDomainLookupByName(conn, name);
+
+  if(c_retval == NULL) {
+    status = napi_get_null(env, &n_retval);
+    assert(status == napi_ok);
+    return n_retval;
+  }
+
+  status = napi_create_external(env, c_retval, NULL, NULL, &n_retval);
+  assert(status == napi_ok);
+
+  return n_retval;
+}
+
+napi_value libvirt_virDomainGetID(napi_env env, napi_callback_info info) {
+    napi_status status;
+    napi_value n_retval;
+    napi_valuetype valuetype;
+
+    size_t argc = 1;
+    napi_value args[1];
+    status = napi_get_cb_info(env, info, &argc, args, NULL, NULL);
+    assert(status == napi_ok);
+
+    if (argc != 1) {
+      napi_throw_type_error(env, NULL, "Wrong number of arguments");
```

Comment:
This is a good point, since we define our JS interface separately we're always passing the right number of arguments so we might not need this code (but we could keep it for safety?).

Alternatively we could forward all arguments from the client
```
open(...args) {
  args[0] = args[0] || 'default'; // this can be done differently
 libvirt_open(...args);
}
but I can't think of a scenario where it would be more beneficial.

the js way is usually to ignore any additional arguments without erroring so i prefer the former

Reply:
Well, let's keep it and see if we ran into some problems with this. We might need to do something with the variables in the glue code.